keps/sig-node/2008-checkpoint-restore/README.md

```diff
@@ -0,0 +1,328 @@
+environments showing that it is useful and production ready.
+
+### Goals
+
+The goal of this KEP is to introduce *checkpoint* and *restore* to the CRI API.
+
+This includes extending the *kubelet* to support checkpointing single containers
+with the User Story 1 in mind (forensic use case).
+
+### Non-Goals
+
+Out of scope of this KEP are high level discussions about how to implement
+container migration. This is only about low level primitives to add
+*checkpoint* and *restore* to the API.
+
+## Proposal
+
+### Implementation
+
+#### Phase 1
+
+One of the first goals is to implement User Story 1. For this forensic
+use case we want to offer the possibility to checkpoint a container
+out of a running Pod without stopping the checkpointed container and without
+letting the container know that it was checkpointed.
+
+The corresponding code changes for the forensic use case can be found in the
+following pull request:
+
+* https://github.com/kubernetes/kubernetes/pull/104907
+
+The goal is to introduce *checkpoint* and *restore* in a bottom-up approach.
+In a first step we only want to extend the CRI API to trigger a checkpoint
+by the container engine and to have the low level primitives in the *kubelet*
+to trigger a checkpoint. It is necessary to enable the feature gate
+`ContainerCheckpointRestore` to be able to checkpoint containers.
+
+In the corresponding pull request a checkpoint is triggered using the *kubelet*
+API:
+
+```
+curl -skv -X POST "https://localhost:10250/checkpoint/default/counters/wildfly"
+```
+
+For the first implementation we do not want to support restore in the
+*kubelet*. With the focus on the forensic use case the restore should happen
+outside of Kubernetes. The restore is a container engine only operation
+in this first step.
+
+The forensic use case is targeted to be part of the next (1.23) release.
+
+### User Stories
+
+#### Story 1
+
+To analyze unusual activities in a container, the container should
+be checkpointed without stopping the container or without the container
+knowing it was checkpointed. Using checkpointing it is possible to take
+a copy of a running container for forensic analysis. The container will
+continue to run without knowing a copy was created. This copy can then
+be restored in another (sandboxed) environment in the context of another
+container engine for detailed analysis of a possible attack.
+
+#### Story 2
+
+Although containers are supposed to be stateless there are still containers
+which either require some time to startup or which have data cached.
+If a system is rebooted to update the kernel these containers can be
+checkpointed. Once the system has been rebooted into the new kernel the
+container can be restored and continue from its previous memory state
+without the need to wait for long start up times or to reload cached data
+from disk.
+
+#### Story 3
+
+Just as in Story 2 a stateful container needs to be moved to another pod.
```

Comment:
This probably doesn't belong in this kep, because it is impossible to address whether the current approach solves this problem without a design of "what does its mean to move a container across pods" and "how would a user indicate that intent is realized".  Maybe I'm just missing where we define what this means?

All I see for this is the reference to: `See Proposal and corresponding implementation pull requests.` which doesn't explain what is required for this story.

Reply:
Are you saying to remove the additional possible use cases of checkpoint/restore.